File: sdk/python/alcedo_sdk/resources/policies.py
from typing import Any, Optional

from .base import BaseResource


class PoliciesResource(BaseResource):
    async def list(self) -> dict:
        return await self._request("GET", "/api/policies")

    async def get(self, id: str) -> dict:
        return await self._request("GET", f"/api/policies/{id}")

    async def create(self, data: dict) -> dict:
        return await self._request("POST", "/api/policies", json=data)

    async def update(self, id: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/policies/{id}", json=data)

    async def delete(self, id: str) -> dict:
        return await self._request("DELETE", f"/api/policies/{id}")

    async def list_permissions(self, id: str) -> dict:
        return await self._request("GET", f"/api/policies/{id}/permissions")

    async def create_permission(self, id: str, data: dict) -> dict:
        return await self._request("POST", f"/api/policies/{id}/permissions", json=data)

    async def update_permission(self, id: str, permission_id: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/policies/{id}/permissions/{permission_id}", json=data)

    async def delete_permission(self, id: str, permission_id: str) -> dict:
        return await self._request("DELETE", f"/api/policies/{id}/permissions/{permission_id}")

    async def list_plugins(self, id: str) -> dict:
        return await self._request("GET", f"/api/policies/{id}/plugins")

    async def list_plugin_policies(self, plugin_slug: str) -> dict:
        return await self._request("GET", f"/api/policies/plugins/{plugin_slug}")

    async def assign_plugin_policy(self, plugin_slug: str, policy_id: str) -> dict:
        return await self._request("POST", f"/api/policies/plugins/{plugin_slug}/policies", json={"policy_id": policy_id})

    async def unassign_plugin_policy(self, plugin_slug: str, policy_id: str) -> dict:
        return await self._request("DELETE", f"/api/policies/plugins/{plugin_slug}/policies/{policy_id}")

File: sdk/alcedocore-sdk-node/src/policies.ts
export function createPoliciesResource(ky: any) {
  return {
    list: (options?: any) => ky.get("app/policies", options).json(),
    get: (id: string, options?: any) => ky.get(`app/policies/${encodeURIComponent(id)}`, options).json(),
    create: (data: any, options?: any) => ky.post("app/policies", { json: data, ...options }).json(),
    update: (id: string, data: any, options?: any) => ky.put(`app/policies/${encodeURIComponent(id)}`, { json: data, ...options }).json(),
    delete: (id: string, options?: any) => ky.delete(`app/policies/${encodeURIComponent(id)}`, options).json(),
    listPermissions: (id: string, options?: any) => ky.get(`app/policies/${encodeURIComponent(id)}/permissions`, options).json(),
    createPermission: (id: string, data: any, options?: any) => ky.post(`app/policies/${encodeURIComponent(id)}/permissions`, { json: data, ...options }).json(),
    updatePermission: (id: string, permissionId: string, data: any, options?: any) => ky.put(`app/policies/${encodeURIComponent(id)}/permissions/${encodeURIComponent(permissionId)}`, { json: data, ...options }).json(),
    deletePermission: (id: string, permissionId: string, options?: any) => ky.delete(`app/policies/${encodeURIComponent(id)}/permissions/${encodeURIComponent(permissionId)}`, options).json(),
    deleteCollectionPermissions: (id: string, collectionName: string, options?: any) => ky.delete(`app/policies/${encodeURIComponent(id)}/permissions/collection/${encodeURIComponent(collectionName)}`, options).json(),
    listAssignedPlugins: (policyId: string, options?: any) => ky.get(`app/policies/${encodeURIComponent(policyId)}/plugins`, options).json(),
    listPluginPolicies: (pluginSlug: string, options?: any) => ky.get(`plugins/${encodeURIComponent(pluginSlug)}/policies`, options).json(),
    assignPluginPolicy: (pluginSlug: string, policyId: string, options?: any) => ky.post(`plugins/${encodeURIComponent(pluginSlug)}/policies`, { json: { policy_id: policyId }, ...options }).json(),
    unassignPluginPolicy: (pluginSlug: string, policyId: string, options?: any) => ky.delete(`plugins/${encodeURIComponent(pluginSlug)}/policies/${encodeURIComponent(policyId)}`, options).json(),
  };
}
